commit 04283a4e02ff12fcb4139d29de3380774601e253
Author: Andre Lamurias <a.lamurias@fct.unl.pt>
Date:   Thu Nov 2 14:17:52 2023 +0000

    fix issue #32 and #33

diff --git a/src/graphmb/main.py b/src/graphmb/main.py
index f5323af..132dd27 100644
--- a/src/graphmb/main.py
+++ b/src/graphmb/main.py
@@ -496,7 +496,7 @@ def main():
             old_args = copy.deepcopy(args)
             args.graph_alpha = 0 # do not use edges 
             args.outname = "ccvae"
-            vae_embs, _ = train_ccvae.run_model_ccvae(dataset, args, logger, 0,
+            vae_embs, _, _ = train_ccvae.run_model_ccvae(dataset, args, logger, 0,
                                                       use_gnn=False, epochs=args.vaepretrain,
                                                       target_metric=target_metric)
             logger.info("===================================================")
@@ -566,7 +566,6 @@ def main():
             dataset.node_to_label,
             seed=args.seed,
         )
-        breakpoint()
         if args.writebins:
             cluster_to_contig = {k: [] for k in set(contig_labels)}
             for contig, label in enumerate(contig_labels):
